pkg/core/blockchain.go

```diff
@@ -628,6 +629,122 @@ func (bc *Blockchain) jumpToStateInternal(p uint32, stage stateJumpStage) error
 	return nil
 }
 
+// Reset resets chain state to the specified height if possible.
+func (bc *Blockchain) Reset(height uint32) error {
+	currHeight := bc.BlockHeight()
+	if height > currHeight {
+		return fmt.Errorf("current block height is %d, can't reset state to height %d", currHeight, height)
+	}
+	if height == currHeight {
+		return nil
+	}
+	if bc.config.KeepOnlyLatestState {
+		return fmt.Errorf("KeepOnlyLatestState is enabled, state for height %d is outdated and removed from the storage", height)
+	}
+	if bc.config.RemoveUntraceableBlocks && currHeight >= bc.config.MaxTraceableBlocks {
+		return fmt.Errorf("RemoveUntraceableBlocks is enabled, a necessary batch of traceable blocks has already been removed")
+	}
+
+	// Retrieve necessary state before the DB modification.
+	b, err := bc.GetBlock(bc.headerHashes[height])
+	if err != nil {
+		return fmt.Errorf("failed to retrieve block %d: %w", height, err)
+	}
+	sr, err := bc.stateRoot.GetStateRoot(height)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve stateroot for height %d: %w", height, err)
+	}
+	v := bc.dao.Version
+
+	// Remove blocks/transactions/aers from currHeight down to height (not including height itself).
+	hHeight := bc.HeaderHeight()
+	cache := bc.dao.GetPrivate()
+	for i := height + 1; i <= currHeight; i++ {
+		err := cache.DeleteBlock(bc.headerHashes[i])
+		if err != nil {
+			return fmt.Errorf("error while removing block %d: %w", i, err)
+		}
+	}
+	for i := height + 1; i <= hHeight; i++ {
+		cache.DeleteHeader(bc.headerHashes[i])
```

Comment:
Is it needed? We `DeleteBlock` above anyway and if it's just a header it'll be deleted already.

Reply:
Yes, we need, because `DeleteBlock` deletes block and stores just a header instead: https://github.com/nspcc-dev/neo-go/blob/57ec67b3750c70d270c2ddd3983f017223e8969f/pkg/core/dao/dao.go#L749

Reply:
Maybe it's easier to add some `PurgeBlock` then and avoid additional calls?

Reply:
Yes, implemented.